clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp

```diff
@@ -0,0 +1,214 @@
+const char CountingEnumPrefixesOptionName[] = "CountingEnumPrefixes";
+const char CountingEnumSuffixesOptionName[] = "CountingEnumSuffixes";
+
+TaggedUnionMemberCountCheck::TaggedUnionMemberCountCheck(
+    StringRef Name, ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context),
+      StrictMode(Options.get(StrictModeOptionName, false)),
+      EnableCountingEnumHeuristic(
+          Options.get(EnableCountingEnumHeuristicOptionName, true)),
+      RawCountingEnumPrefixes(Options.get(CountingEnumPrefixesOptionName, "")),
+      RawCountingEnumSuffixes(
+          Options.get(CountingEnumSuffixesOptionName, "count")),
+      ParsedCountingEnumPrefixes(
+          utils::options::parseStringList(RawCountingEnumPrefixes)),
+      ParsedCountingEnumSuffixes(
+          utils::options::parseStringList(RawCountingEnumSuffixes)),
+      CountingEnumPrefixesSet(
+          Options.get(CountingEnumPrefixesOptionName).has_value()),
+      CountingEnumSuffixesSet(
+          Options.get(CountingEnumSuffixesOptionName).has_value()),
+      CountingEnumConstantDecl(nullptr) {
+  if (!EnableCountingEnumHeuristic) {
+    if (CountingEnumPrefixesSet)
+      configurationDiag("%0: Counting enum heuristic is disabled but "
+                        "CountingEnumPrefixes is set")
+          << Name;
+    if (CountingEnumSuffixesSet)
+      configurationDiag("%0: Counting enum heuristic is disabled but "
+                        "CountingEnumSuffixes is set")
+          << Name;
+  }
+}
+
+void TaggedUnionMemberCountCheck::storeOptions(
+    ClangTidyOptions::OptionMap &Opts) {
+  Options.store(Opts, StrictModeOptionName, StrictMode);
+  Options.store(Opts, EnableCountingEnumHeuristicOptionName,
+                EnableCountingEnumHeuristic);
+  Options.store(Opts, CountingEnumPrefixesOptionName, RawCountingEnumPrefixes);
+  Options.store(Opts, CountingEnumSuffixesOptionName, RawCountingEnumSuffixes);
+}
+
+void TaggedUnionMemberCountCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      recordDecl(
+          anyOf(isStruct(), isClass()),
+          has(fieldDecl(hasType(qualType(hasCanonicalType(recordType()))))
+                  .bind("union")),
+          has(fieldDecl(hasType(qualType(hasCanonicalType(enumType()))))
+                  .bind("tags")))
+          .bind("root"),
+      this);
+}
+
+static bool isUnion(const FieldDecl *R) {
+  return R->getType().getCanonicalType().getTypePtr()->isUnionType();
+}
+
+static bool isEnum(const FieldDecl *R) {
+  return R->getType().getCanonicalType().getTypePtr()->isEnumeralType();
+}
+
+static bool hasMultipleUnionsOrEnums(const RecordDecl *Rec) {
+  return llvm::count_if(Rec->fields(), isUnion) > 1 ||
+         llvm::count_if(Rec->fields(), isEnum) > 1;
+}
+
+static bool signEquals(const llvm::APSInt &A, const llvm::APSInt &B) {
+  return (A.isNegative() && B.isNegative()) ||
+         (A.isStrictlyPositive() && B.isStrictlyPositive()) ||
+         (A.isZero() && B.isZero());
+}
```

Comment:
**If** the following is equivalent (I would guess `isNegative`/`isStrictlyPositive`/`isZero` are all mutually exclusive properties) then perhaps expressing the query like this would better reflect that we are checking for some *equality* here:

```suggestion
static bool signEquals(const llvm::APSInt &A, const llvm::APSInt &B) {
  return (A.isNegative() == B.isNegative()) &&
         (A.isStrictlyPositive() == B.isStrictlyPositive()) &&
         (A.isZero() == B.isZero());
}
```

-----

Also, I am **very** surprised that `APInt`/`APSInt` does not have the facility for querying the [sign function](https://en.wikipedia.org/wiki/Sign_function) as a single value. Tangentially, perhaps you could add that in a separate patch with a few tests, a function that returns a 3-way enum `NEGATIVE, ZERO, POSITIVE`, and then this function simplifies to `A.getSign() == B.getSign()`, which is the most trivial expansion of the function's name.